### ISSUE TITLE
Add a way to access parent or global data while validating nested resource or associated model

### ISSUE THREAD

**Issue by doshu** [cakephp/cakephp#17705]
### Description

If a nested resource or an associated entity has a validation logic that depends on the parent data or a value of the parent models, there is no way to access that data.

For example, when add a nested validator, the context object contains only the data of the nested resource.
Is there a way to add to the context object the complete data array in the case of a modelless form or the main entity where the validation has been started in the case on a table?

**Comment by doshu** [comment]
is there any update or a plan to implement some features?
unfortunately today is really hard to implement a validation logic that include nested entity where the nested entity data depends on the parent entity data.
this kind of validations are really common in complex softwares

**Comment by markstory** [comment]
> is there any update or a plan to implement some features?

Not currently.

How would you expect to access the parent entity context/data? How would that work if there are multiple parents stacked together? Can you only access the immediate parent or all parents in the chain?

**Comment by doshu** [comment]
parent data should be accessed in the chain. a _parent key should be added to the data in the context recursively. 
a nested validator can access parent data and if there is another level of nested, the second level of data can be accessed too. Nested validator are coupled by design and everything will work fine.

**Comment by markstory** [comment]
Sure, but if you were to write code to do that, what would that code look like?

**Comment by doshu** [comment]
I haven't done a code analysis yet, but probably can be enough to add a third params to te validate method of the validator class that accept the parent context. 
In the validate method then, the parent context could be pushed in the current context under the _parent key.
when running the rule of the nested validator, the current context will be passed as the third param.

whit this solution every nested validator will have the _parent key in its context recurvively until the parent context is null (root validator context)

**Comment by doshu** [comment]
the validate method prototype
public function validate(array $data, bool $newRecord = true, array $parentContext = null): array

when calling in the addNested or addNestedMany callback
$validator->validate($row, $context['newRecord'], $context);

**Comment by doshu** [comment]
Any plan to implement this? Currently i'm working on a big project using laravel and sometimes i try to translate complex validations in the cakephp env, and currently is really difficult or impossible to implement complex validation rules where some conditions depends on parent data in nested validation.

**Comment by markstory** [comment]
> Any plan to implement this

I'm not aware of any plans, do you have an interest in putting a pull request together?

**Comment by doshu** [comment]
i have made a pull requesti, but has not been accepted because of an edit on a protected method. any plan to merge in a next release?

**Comment by dereuromark** [comment]
https://github.com/cakephp/cakephp/pull/18850

**Comment by doshu** [comment]
Thanks, this will be very useful and adapt the functionality to other frameworks. can't wait for a mention for the contribute, if possible😁

**Comment by dereuromark** [comment]
Yes of course. We will add it.
Can you double check the PR? If it resolved your issue fully?

**Comment by doshu** [comment]
yes it does. with this implentation is possible to recursively extract data from the parent context in the case of nested validation, and is also possible to know what is the index of the resource under validation

**Comment by doshu** [comment]
The feature has not be mentioned in the 5.3 rc1 news and documentation.
Sad to not see any mention for the PR

**Comment by dereuromark** [comment]
@doshu I agree. We should mention the powerful additions to validation functionality.

**Comment by markstory** [comment]
There were never any docs written for this pull request, so it didn't end up in the release notes either.
